File: java/63-unique-paths-ii.java
class Solution {
    int[][] obstacleGrid;
    int[][] dp;

    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        this.obstacleGrid = obstacleGrid;
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;

        dp = new int[m][n];
        dp[m - 1][n - 1] = 1;

        return dfs(0, 0);
    }

    private int dfs(int i, int j) {
        if (i == obstacleGrid.length || j == obstacleGrid[0].length || obstacleGrid[i][j] == 1)
            return 0;
        if (dp[i][j] > 0)
            return dp[i][j];

        return dp[i][j] = dfs(i + 1, j) + dfs(i, j + 1);

    }
}

File: c++/63-unique-paths-ii.cpp
class Solution {
public:
    
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int length = obstacleGrid.size();
        int width = obstacleGrid[0].size();
        
        int last[width+1];
        int llast=1;
        int j;
        vector<int> vec = obstacleGrid[0];
        for (j = 1; j <= width; j++) {
            if (vec[j-1] == 1) {
                if (j==1) {
                    return 0;
                }
                break;
            }
            last[j] = 1;
        }
        memset(last+(j), 0, (width-j+1)*sizeof(int));
        
        for (int i=1; i < length; i++) {
            vec = obstacleGrid[i];
            llast = 0;
            for (int j = 1; j <= width; j++) {
                if (vec[j-1] == 1) {
                    last[j] = 0;
                } else {
                    last[j] = last[j] + llast;
                }

                llast = last[j];
            }
            
        }
        return last[width];
    }
};